server/src/test/java/org/elasticsearch/action/admin/cluster/node/tasks/TestTaskPlugin.java

```diff
@@ -469,4 +492,70 @@ protected UnblockTestTasksRequestBuilder(ElasticsearchClient client,
         }
     }
 
+    private static class OriginAssertingInterceptor implements TransportInterceptor {
+        private final ThreadContext threadContext;
+
+        private OriginAssertingInterceptor(ThreadContext threadContext) {
+            this.threadContext = threadContext;
+        }
+
+        @Override
+        public AsyncSender interceptSender(AsyncSender sender) {
+            return new AsyncSender() {
+                @Override
+                public <T extends TransportResponse> void sendRequest(
+                        Transport.Connection connection, String action, TransportRequest request,
+                        TransportRequestOptions options, TransportResponseHandler<T> handler) {
+                    if (action.startsWith("indices:data/write/bulk[s]")) {
+                        /*
+                         * We can't reason about these requests because
+                         * *sometimes* they should have the origin, if they are
+                         * running on the node that stores the task. But
+                         * sometimes they won't be and in that case they don't
+                         * need the origin. Either way, the interesting work is
+                         * done by checking that the main bulk request
+                         * (without the [s] part) has the origin.
+                         */
+                        sender.sendRequest(connection, action, request, options, handler);
+                        return;
+                    }
+                    String expectedOrigin = shouldHaveOrigin(action, request) ? TASKS_ORIGIN : null;
```

Comment:
I'd probably go with the following since we'll probably consider adding other origins within server:

```java
if (shouldHaveOrigin(action, request)) {
    String actualOrigin = threadContext.getTransient(ThreadContext.ACTION_ORIGIN_TRANSIENT_NAME);
    if (Objects.equals(TASKS_ORIGIN, actualOrigin)) {
        sender.sendRequest(connection, action, request, options, handler);
    } else {
        handler.handleException(new TransportException("should have origin of ["
                            + TASKS_ORIGIN + "] but was [" + actualOrigin + "] action was ["
                            + action + "][" + request + "]"));
    }
} else {
    sender.sendRequest(connection, action, request, options, handler);
}
```

Reply:
The assertion I have is pretty tight - if the request shouldn't have the `origin` set then I assert that it is null. I like that! Yours won't assert anything if the request shouldn't have an origin on it. Like, the request could have an origin and my test won't notice! I don't think that is over specifying this here because the origin seems like a fairly careful concept.

Reply:
That’s fair